cassandra/connection.py

```diff
@@ -732,3 +743,96 @@ def deliver(self, timeout=None):
             raise OperationTimedOut()
         else:
             return self.responses
+
+
+class HeartbeatFuture(object):
+    def __init__(self, connection, owner):
+        self._exception = None
+        self._event = Event()
+        self.connection = connection
+        self.owner = owner
+        log.debug("Sending options message heartbeat on idle connection %s %s",
+                  id(connection), connection.host)
+        with connection.lock:
+            if connection.in_flight < connection.max_request_id:
+                connection.in_flight += 1
+                connection.send_msg(OptionsMessage(), connection.get_request_id(), self._options_callback)
+            else:
+                self._exception = Exception("Failed to send heartbeat because connection 'in_flight' exceeds threshold")
```

Comment:
Considering that we only heartbeat if a connection appears idle, arriving here and finding no request id would indicate that there are max_request_id requests out at this very instant, while no responses have been seen on this connection for the last idle_heartbeat_interval seconds.
I think in a healthy, high load situation we would not be here. I can imagine wanting to wait is if the idle_heartbeat_interval is < normal request latency, but that would be a silly configuration.